shared/subspace-data-retrieval/src/segment_fetcher.rs

```diff
@@ -14,5 +14,168 @@
+use subspace_archiving::archiver::Segment;
+use subspace_archiving::reconstructor::{Reconstructor, ReconstructorError};
+use subspace_core_primitives::{
+    ArchivedHistorySegment, Piece, RecordedHistorySegment, SegmentIndex,
+};
+use subspace_erasure_coding::ErasureCoding;
+use tokio::sync::Semaphore;
+use tokio::task::spawn_blocking;
+use tracing::{debug, trace, warn};
+
+/// Segment getter errors.
+#[derive(Debug, thiserror::Error)]
+pub enum SegmentGetterError {
+    /// Piece getter error
+    #[error("Failed to get enough segment pieces")]
+    PieceGetter { segment_index: SegmentIndex },
+
+    /// Segment reconstruction error
+    #[error("Segment reconstruction error: {source:?}")]
+    SegmentReconstruction {
+        #[from]
+        source: ReconstructorError,
+    },
+
+    /// Segment decoding error
+    #[error("Segment data decoding error: {source:?}")]
+    SegmentDecoding {
+        #[from]
+        source: parity_scale_codec::Error,
+    },
+}
+
+/// Concurrently downloads the pieces for `segment_index`, and reconstructs the segment.
+pub async fn download_segment<PG>(
+    segment_index: SegmentIndex,
+    piece_getter: &PG,
+    erasure_coding: ErasureCoding,
+) -> Result<Segment, SegmentGetterError>
+where
+    PG: ObjectPieceGetter,
+{
+    let reconstructor = Reconstructor::new(erasure_coding);
+
+    let segment_pieces = download_segment_pieces(segment_index, piece_getter).await?;
+
+    let segment = spawn_blocking(move || reconstructor.reconstruct_segment(&segment_pieces))
+        .await
+        .expect("Panic if blocking task panicked")?;
+
+    Ok(segment)
+}
+
+/// Concurrently downloads the pieces for `segment_index`.
+// This code was copied and modified from subspace_service::sync_from_dsn::download_and_reconstruct_blocks():
+// <https://github.com/autonomys/subspace/blob/d71ca47e45e1b53cd2e472413caa23472a91cd74/crates/subspace-service/src/sync_from_dsn/import_blocks.rs#L236-L322>
+//
+// TODO: pass a lower concurrency limit into this function, to avoid overwhelming residential routers or slow connections
+pub async fn download_segment_pieces<PG>(
+    segment_index: SegmentIndex,
+    piece_getter: &PG,
+) -> Result<Vec<Option<Piece>>, SegmentGetterError>
+where
+    PG: ObjectPieceGetter,
+{
+    debug!(%segment_index, "Retrieving pieces of the segment");
+
+    let semaphore = &Semaphore::new(RecordedHistorySegment::NUM_RAW_RECORDS);
+
+    let mut received_segment_pieces = segment_index
+        .segment_piece_indexes_source_first()
+        .into_iter()
+        .map(|piece_index| {
+            // Source pieces will acquire permit here right away
+            let maybe_permit = semaphore.try_acquire().ok();
+
+            async move {
+                let permit = match maybe_permit {
+                    Some(permit) => permit,
+                    None => {
+                        // Other pieces will acquire permit here instead
+                        match semaphore.acquire().await {
```

Comment:
I like `async-lock`'s semaphore more because it doesn't support closing and doesn't need to be checked this way